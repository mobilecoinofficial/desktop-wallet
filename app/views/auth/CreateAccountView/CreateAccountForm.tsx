import React, { useState } from 'react';
import type { FC } from 'react';

import { Box, Button, FormHelperText, Typography } from '@material-ui/core';
import { Formik, Form, Field } from 'formik';
import type { FormikHelpers } from 'formik';
import { Checkbox, TextField } from 'formik-material-ui';
import { useTranslation } from 'react-i18next';
import * as Yup from 'yup';

import { SubmitButton, TermsOfUseDialog } from '../../../components';
import type { FullServiceContextValue } from '../../../contexts/FullServiceContext';
import useFullService from '../../../hooks/useFullService';
import useIsMountedRef from '../../../hooks/useIsMountedRef';

export interface CreateAccountFormValues {
  accountName: string;
  checkedTerms: boolean;
  password: string;
  passwordConfirmation: string;
  submit: null;
}

interface CreateAccountFormPseudoProps {
  isMountedRef: { current: boolean };
  createAccount: FullServiceContextValue['createAccount'];
}

export const createAccountFormOnSubmit = async (
  pseudoProps: CreateAccountFormPseudoProps,
  values: CreateAccountFormValues,
  helpers: FormikHelpers<CreateAccountFormValues>
): Promise<void> => {
  const { isMountedRef, createAccount } = pseudoProps;
  const { accountName, password } = values;
  const { setStatus, setErrors, setSubmitting } = helpers;
  setSubmitting(true);

  try {
    await createAccount(accountName, password);

    if (isMountedRef.current) {
      setStatus({ success: true });
      setSubmitting(false);
    }
  } catch (err) {
    if (isMountedRef.current) {
      setStatus({ success: false });
      setErrors({ submit: err.message });
      setSubmitting(false);
    }
  }
};

interface CreateAccountFormProps {
  isTest: boolean | undefined;
  onSubmit: typeof createAccountFormOnSubmit;
}

const CreateAccountForm: FC<CreateAccountFormProps> = ({
  isTest,
  onSubmit,
}: CreateAccountFormProps) => {
  const isMountedRef = useIsMountedRef();
  const { t } = useTranslation('CreateAccountForm');
  const { createAccount } = useFullService();

  const [canCheck, setCanCheck] = useState(false);
  const [open, setOpen] = useState(false);

  const handleCloseTerms = () => {
    setCanCheck(true);
    setOpen(false);
  };

  // FIX-ME: This hack is to avoid opening the Dialog -- which is causing some
  // headaches in testing.
  const handleClickOpen = () => (isTest ? handleCloseTerms() : setOpen(true));

  const handleOnSubmit = async (
    values: CreateAccountFormValues,
    helpers: FormikHelpers<CreateAccountFormValues>
  ) => {
    const pseduoProps = { createAccount, isMountedRef };
    await onSubmit(pseduoProps, values, helpers);
  };

  const initialValues = {
    accountName: '',
    checkedTerms: false,
    password: '',
    passwordConfirmation: '',
    submit: null,
  };

  const validationSchema = Yup.object().shape({
    accountName: Yup.string().max(64, t('accountNameValidation')),
    // CBB: It appears that the checkedTerms error message is not working properly.
    checkedTerms: Yup.bool().oneOf([true], t('checkedTermsValidation')),
    password: Yup.string()
      .min(8, t('passwordMin'))
      .max(99, t('passwordMax'))
      .required(t('passwordRequired')),
    passwordConfirmation: Yup.string()
      .oneOf([Yup.ref('password')], t('passwordConfirmationRef'))
      .required(t('passwordConfirmationRequired')),
  });

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={handleOnSubmit}
      validationSchema={validationSchema}
    >
      {({ errors, isSubmitting, dirty, isValid, submitForm }) => (
        <Form name="CreateAccountFormName">
          <Field
            id="CreateAccountForm-accountNameField"
            component={TextField}
            fullWidth
            label={t('nameLabel')}
            name="accountName"
          />
          <Field
            id="CreateAccountForm-passwordField"
            component={TextField}
            fullWidth
            label={t('passwordLabel')}
            margin="dense"
            name="password"
            type="password"
          />
          <Field
            id="CreateAccountForm-passwordConfirmationField"
            component={TextField}
            fullWidth
            label={t('passwordConfirmationLabel')}
            margin="dense"
            name="passwordConfirmation"
            type="password"
          />
          <Box pt={1} display="flex">
            <Box display="flex" alignItems="center" flexDirection="row-reverse">
              <Box>
                <Typography display="inline">{t('acceptTerms')}</Typography>
                <Button color="primary" onClick={handleClickOpen}>
                  {t('acceptTermsButton')}
                </Button>
              </Box>
              <Field
                component={Checkbox}
                type="checkbox"
                name="checkedTerms"
                disabled={!canCheck}
                indeterminate={!canCheck}
              />
            </Box>
          </Box>
          {!canCheck && <FormHelperText focused>{t('acceptTermsFormHelper')}</FormHelperText>}
          {errors.submit && (
            <Box mt={3}>
              <FormHelperText error>{errors.submit}</FormHelperText>
            </Box>
          )}
          <SubmitButton
            disabled={!dirty || !isValid || isSubmitting}
            onClick={submitForm}
            isSubmitting={isSubmitting}
          >
            {t('createAccountButton')}
          </SubmitButton>
          <TermsOfUseDialog open={open} handleCloseTerms={handleCloseTerms} />
        </Form>
      )}
    </Formik>
  );
};

export default CreateAccountForm;
