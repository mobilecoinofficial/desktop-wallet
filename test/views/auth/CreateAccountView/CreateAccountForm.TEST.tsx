import React from 'react';

import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import CreateAccountForm, {
  createAccountFormOnSubmit,
} from '../../../../app/views/auth/CreateAccountView/CreateAccountForm';
import renderSnapshot from '../../../renderSnapshot';

jest.mock('../../../../app/hooks/useFullService');

function setupComponent() {
  // Variables
  const validAccountName64 = '64llllllllllllllllllllllllllllllllllllllllllllllllllllllllllllll';
  const invalidAccountName65 = '65lllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllll';
  const invalidPasswordShort = 'shooort';
  const validPassword99 =
    'longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglon';

  const { mockUseFullServiceValues } = renderSnapshot(
    <CreateAccountForm onSubmit={createAccountFormOnSubmit} isTest />
  );

  // Render Elements
  const form = screen.getByRole('form');
  const accountNameField = screen.getByLabelText('Account Name (optional)', {
    selector: 'input',
  }) as HTMLInputElement;
  const passwordField = screen.getByLabelText('Password', {
    selector: 'input',
  }) as HTMLInputElement;
  const passwordConfirmationField = screen.getByLabelText('Password Confirmation', {
    selector: 'input',
  }) as HTMLInputElement;
  const checkTermsField = screen.getByRole('checkbox') as HTMLInputElement;
  const termsButton = screen.getByRole('button', {
    name: 'Terms of Use',
  });
  const submitButton = screen.getByRole('button', { name: 'Create Account' });

  return {
    accountNameField,
    checkTermsField,
    form,
    invalidAccountName65,
    invalidPasswordShort,
    mockUseFullServiceValues,
    passwordConfirmationField,
    passwordField,
    submitButton,
    termsButton,
    validAccountName64,
    validPassword99,
  };
}

function setupOnSubmit() {
  // Mocks
  const mockSetStatus = jest.fn();
  const mockSetSubmitting = jest.fn();
  const mockSetErrors = jest.fn();
  const mockCreateAccount = jest.fn();

  // Variables
  const accountName = 'account name';
  const password = 'password';
  const isMountedRefTrue = { current: true };
  const isMountedRefFalse = { current: false };
  const helpers = {
    setErrors: mockSetErrors,
    setStatus: mockSetStatus,
    setSubmitting: mockSetSubmitting,
  };

  return {
    accountName,
    helpers,
    isMountedRefFalse,
    isMountedRefTrue,
    mockCreateAccount,
    mockSetErrors,
    mockSetStatus,
    mockSetSubmitting,
    password,
  };
}

describe('CreateAccountForm', () => {
  describe('component', () => {
    describe('initalValues', () => {
      test('sets correct initial values', async () => {
        const { form } = setupComponent();
        const expectedInitialValues = {
          accountName: '',
          checkedTerms: false,
          password: '',
          passwordConfirmation: '',
        };

        expect(form).toHaveFormValues(expectedInitialValues);
      });
    });

    describe('validations', () => {
      test('limits account name to 64 characters.', async () => {
        const { accountNameField, invalidAccountName65, validAccountName64 } = setupComponent();
        const expectedErrorMessage = 'Account Name cannot be more than 64 characters.';

        // Fill out with too long account name
        userEvent.type(accountNameField, invalidAccountName65);
        userEvent.tab(); // Tab to trigger validations

        // Await because validations are async
        const errorMessage = await screen.findByText(expectedErrorMessage);
        expect(errorMessage).toBeInTheDocument();

        // Clear and use name under the limit
        userEvent.clear(accountNameField);
        userEvent.type(accountNameField, validAccountName64);
        userEvent.tab(); // Tab to trigger validations
        await waitFor(() => {
          expect(errorMessage).not.toBeInTheDocument();
        });
      });

      test('checkbox is disabled until reading terms', async () => {
        const { checkTermsField, termsButton } = setupComponent();
        const expectedTermsMessage = 'You must read the Terms of Use before using the wallet.';

        // The checkbox is diabled until user has read terms
        expect(checkTermsField.value).toBe('false');
        userEvent.click(checkTermsField);
        expect(checkTermsField.value).toBe('false');

        const termsMessage = await screen.findByText(expectedTermsMessage);
        expect(termsMessage).toBeInTheDocument();

        // Reading the terms removes message and allows you to click terms
        userEvent.click(termsButton);
        await waitFor(() => {
          expect(termsMessage).not.toBeInTheDocument();
        });
        userEvent.click(checkTermsField);
        expect(checkTermsField.value).toBe('true');
      });

      test('password is required and must be between 8 and 99 characters', async () => {
        const { passwordField, validPassword99, invalidPasswordShort } = setupComponent();
        const expectedShortErrorMessage = 'Password must be at least 8 characters in length.';
        const expectedRequiredErrorMessage = 'Password is required';
        const expectedLongErrorMessage = 'Passwords cannot be more than 99 characters.';

        // Type up to 1 short from valid and check error
        userEvent.type(passwordField, invalidPasswordShort);
        userEvent.tab(); // Tab to trigger validations

        // Await because validations are async
        const shortErrorMessage = await screen.findByText(expectedShortErrorMessage);
        expect(shortErrorMessage).toBeInTheDocument();

        // Add a character to become valid
        userEvent.type(passwordField, '1');
        userEvent.tab(); // Tab to trigger validations
        await waitFor(() => {
          expect(shortErrorMessage).not.toBeInTheDocument();
        });

        // Clear to show required error
        userEvent.clear(passwordField);
        // Await because validations are async
        const requiredErrorMessage = await screen.findByText(expectedRequiredErrorMessage);
        expect(requiredErrorMessage).toBeInTheDocument();
        expect(shortErrorMessage).not.toBeInTheDocument();

        // Write a password at maximum valid length + 1
        userEvent.type(passwordField, validPassword99);
        userEvent.type(passwordField, '1');
        userEvent.tab(); // Tab to trigger validations

        // Await because validations are async
        const longErrorMessage = await screen.findByText(expectedLongErrorMessage);
        expect(longErrorMessage).toBeInTheDocument();
        expect(shortErrorMessage).not.toBeInTheDocument();
        expect(requiredErrorMessage).not.toBeInTheDocument();

        // Finally, backspace to become valid again
        userEvent.type(passwordField, '{backspace}1');
        await waitFor(() => {
          expect(longErrorMessage).not.toBeInTheDocument();
          expect(shortErrorMessage).not.toBeInTheDocument();
          expect(requiredErrorMessage).not.toBeInTheDocument();
        });
      });

      test('password confirmation is required and must match password', async () => {
        const { passwordConfirmationField, passwordField, validPassword99 } = setupComponent();
        const expectedMustMatchMessage = 'Must match Password';
        const expectedRequiredErrorMessage = 'Password Confirmation is required';

        // Type different passwords and password confirmations
        userEvent.type(passwordField, validPassword99);
        userEvent.type(passwordConfirmationField, 'something completely different');
        userEvent.tab(); // Tab to trigger validations
        // Await because validations are async

        const mustMatchErrorMessage = await screen.findByText(expectedMustMatchMessage);
        expect(mustMatchErrorMessage).toBeInTheDocument();

        // Clear password confirmation to get error
        userEvent.clear(passwordConfirmationField);
        userEvent.tab(); // Tab to trigger validations
        // Await because validations are async

        const requiredErrorMessage = await screen.findByText(expectedRequiredErrorMessage);
        expect(requiredErrorMessage).toBeInTheDocument();

        // Type matching confirmation to dismiss errors
        userEvent.type(passwordConfirmationField, validPassword99);
        userEvent.tab(); // Tab to trigger validations
        await waitFor(() => {
          expect(requiredErrorMessage).not.toBeInTheDocument();
          expect(mustMatchErrorMessage).not.toBeInTheDocument();
        });
      });
    });

    describe('submit', () => {
      // test('calls createAccount hook with a password and accountName', async () => {
      //   const {
      //     accountNameField,
      //     checkTermsField,
      //     mockUseFullServiceValues,
      //     passwordConfirmationField,
      //     passwordField,
      //     termsButton,
      //     submitButton,
      //     validAccountName64,
      //     validPassword99,
      //   } = setupComponent();

      //   // First tests that the button is disabled
      //   expect(submitButton).toBeDisabled();
      //   userEvent.click(submitButton);
      //   await waitFor(() => {
      //     expect(mockUseFullServiceValues.createAccount).not.toBeCalled();
      //   });

      //   // Enter valid form information
      //   userEvent.type(accountNameField, validAccountName64);
      //   userEvent.type(passwordField, validPassword99);
      //   userEvent.type(passwordConfirmationField, validPassword99);
      //   userEvent.click(termsButton);
      //   userEvent.click(checkTermsField);
      //   expect(accountNameField.value).toBe(validAccountName64);
      //   expect(passwordField.value).toBe(validPassword99);
      //   expect(passwordConfirmationField.value).toBe(validPassword99);
      //   expect(checkTermsField.value).toBe('true');

      //   // Submit
      //   await waitFor(() => {
      //     expect(submitButton).not.toBeDisabled();
      //   });
      //   userEvent.click(submitButton);

      //   await waitFor(() => {
      //     expect(mockUseFullServiceValues.createAccount).toBeCalledWith(
      //       validAccountName64,
      //       validPassword99
      //     );
      //   });
      // });

      test('displays error when thrown', async () => {
        const expectedErrorMessage = 'I am an error!';
        const {
          accountNameField,
          checkTermsField,
          mockUseFullServiceValues,
          passwordConfirmationField,
          passwordField,
          termsButton,
          submitButton,
          validAccountName64,
          validPassword99,
        } = setupComponent();
        // @ts-ignore mock
        mockUseFullServiceValues.createAccount.mockImplementation(() => {
          throw new Error(expectedErrorMessage);
        });

        // Enter valid form information & Submit
        userEvent.type(accountNameField, validAccountName64);
        userEvent.type(passwordField, validPassword99);
        userEvent.type(passwordConfirmationField, validPassword99);
        userEvent.click(termsButton);
        userEvent.click(checkTermsField);
        userEvent.click(submitButton);

        await waitFor(() => {
          expect(screen.getByText(expectedErrorMessage)).toBeInTheDocument();
        });
      });
    });
  });

  describe('functions', () => {
    // CBB: I don't like this. But I want to make sure that the correct
    // hooks are being set with the different scenarios.
    describe('createAccountFormOnSubmit', () => {
      test('calls CreateAccount and helpers when mounted', async () => {
        const {
          accountName,
          helpers,
          isMountedRefTrue,
          mockCreateAccount,
          password,
        } = setupOnSubmit();

        const pseudoProps = {
          createAccount: mockCreateAccount,
          isMountedRef: isMountedRefTrue,
        };
        const values = { accountName, password };
        // @ts-ignore mock
        await createAccountFormOnSubmit(pseudoProps, values, helpers);

        expect(mockCreateAccount).toBeCalledWith(accountName, password);
        expect(helpers.setStatus).toBeCalledWith({ success: true });
        expect(helpers.setSubmitting).toBeCalledWith(false);
        expect(helpers.setErrors).not.toBeCalled();
      });

      test('calls CreateAccount and but not helpers when unmounted', async () => {
        const {
          accountName,
          helpers,
          isMountedRefFalse,
          mockCreateAccount,
          password,
        } = setupOnSubmit();

        const pseudoProps = {
          createAccount: mockCreateAccount,
          isMountedRef: isMountedRefFalse,
        };
        const values = { accountName, password };

        // @ts-ignore mock
        await createAccountFormOnSubmit(pseudoProps, values, helpers);

        expect(helpers.setStatus).not.toBeCalled();
        expect(helpers.setSubmitting).not.toBeCalled();
      });

      test('correctly sets helpers when call fails whem mounted', async () => {
        const {
          accountName,
          helpers,
          isMountedRefTrue,
          mockCreateAccount,
          password,
        } = setupOnSubmit();

        const errorMessage = 'error message.';
        mockCreateAccount.mockRejectedValueOnce(new Error(errorMessage));
        const pseudoProps = {
          createAccount: mockCreateAccount,
          isMountedRef: isMountedRefTrue,
        };
        const values = { accountName, password };

        // @ts-ignore mock
        await createAccountFormOnSubmit(pseudoProps, values, helpers);
        expect(mockCreateAccount).toBeCalled();

        expect(helpers.setStatus).toBeCalledWith({ success: false });
        expect(helpers.setSubmitting).toBeCalledWith(false);
        expect(helpers.setErrors).toBeCalledWith({ submit: errorMessage });
      });

      test('does not call helpers when call fails when unmounted', async () => {
        const {
          accountName,
          helpers,
          isMountedRefFalse,
          mockCreateAccount,
          password,
        } = setupOnSubmit();

        const errorMessage = 'error message.';
        mockCreateAccount.mockRejectedValueOnce(new Error(errorMessage));
        const pseudoProps = {
          createAccount: mockCreateAccount,
          isMountedRef: isMountedRefFalse,
        };
        const values = { accountName, password };

        // @ts-ignore mock
        await createAccountFormOnSubmit(pseudoProps, values, helpers);

        expect(helpers.setStatus).not.toBeCalled();
        expect(helpers.setSubmitting).not.toBeCalled();
        expect(helpers.setErrors).not.toBeCalled();
      });
    });
  });
});
