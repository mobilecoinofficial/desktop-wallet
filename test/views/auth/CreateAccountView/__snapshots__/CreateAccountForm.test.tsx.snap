// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAccountForm component render it renders correctly 1`] = `
<DocumentFragment>
  <form
    action="#"
    name="CreateAccountFormName"
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink="false"
        for="accountNameField"
        id="accountNameField-label"
      >
        Account Name (optional)
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input"
          id="accountNameField"
          name="accountName"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense"
        data-shrink="false"
        for="passwordField"
        id="passwordField-label"
      >
        Password
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
          id="passwordField"
          name="password"
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense"
        data-shrink="false"
        for="passwordConfirmationField"
        id="passwordConfirmationField-label"
      >
        Password Confirmation
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
          id="passwordConfirmationField"
          name="passwordConfirmation"
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root"
    >
      <div
        class="MuiBox-root MuiBox-root"
      >
        <div
          class="MuiBox-root MuiBox-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-displayInline"
          >
            I have read and accept the
          </p>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Terms of Use
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <span
          aria-disabled="true"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorSecondary MuiCheckbox-indeterminate PrivateSwitchBase-disabled MuiCheckbox-disabled MuiIconButton-colorSecondary MuiIconButton-disabled MuiButtonBase-disabled"
          tabindex="-1"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input"
              data-indeterminate="true"
              disabled=""
              name="checkedTerms"
              type="checkbox"
              value="false"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-2 10H7v-2h10v2z"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
    <p
      class="MuiFormHelperText-root MuiFormHelperText-focused"
    >
      You must read the Terms of Use before using the wallet.
    </p>
    <div
      class="MuiBox-root MuiBox-root makeStyles-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-containedSizeLarge MuiButton-sizeLarge MuiButton-disabled MuiButton-fullWidth MuiButtonBase-disabled"
        data-testid="submit-button"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Create Account
        </span>
      </button>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        TEST BUTTON
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
